### ISSUE TITLE
getProgress is incorrect in AssetManager

### ISSUE THREAD

**Issue by WalkerWhite** [libgdx/libgdx#6411]
#### Issue details
The method getProgress in AssetManager is capable of returning 1.0 even when the loading queue is not empty.  The problem has to do with how True Type Fonts are loaded.  Loading the .ttf to create the .gen, and loading the .gen to create the Bitmap are separate loading operations.  Apparently only one of them counts against the variable `toLoad`.  However, the gen is apparently counted for `loaded`.  So the code thinks that loading is done, even when there is a non-empty loading queue (the bitmap font dependency) or when peakTasks > 0.

#### Reproduction steps/code

This is a race condition.  It is possible that you will get lucky and getProgress will work.  However, I have a project that loads fonts last, and 100% of the time the final font fails to load.  Either the loadQueue is not empty, or the task it running even though `toLoad == loaded`

#### Version of LibGDX and/or relevant dependencies
1.9.14

#### Stacktrace
N/A

#### Please select the affected platforms
- [x ] Android
- [x ] iOS
- [x ] HTML/GWT
- [x ] Windows
- [x ] Linux
- [x ] MacOS

**Comment by MrStahlfelge** [comment]
Can you give an example code how you load the fonts?

**Comment by WalkerWhite** [comment]
I have a project that demonstrates it here:

https://www.cs.cornell.edu/courses/cs3152/2021sp/materials/demos/Assets.zip

I have created an AssetDirectory class, which is a subclass of AssetManager.  This allows you to specify all of your assets in a JSON for data-driven loading like this:

```
{
    "textures": {
        "background": "images/background.png",
        "beam": {
            "file":     "images/beam.png",
            "minFilter": "linear",
            "magFilter": "nearest"
        }    
    },
    "fonts": {
        "times": {
            "file": "fonts/LucidaGrande.ttf",
            "size": 16
        },
        "times": {
            "file": "fonts/TimesRoman.ttf",
            "size": 32
        }
    }
}
```

The details of how this work are not that important (it uses a pluggable architecture like AssetManager to attach JSON parsers in addition to loaders).  But it reads the JSON file synchronously first and then queues up all of the contents asynchronously.  Right now, everything is loaded in order, so the fonts are loaded last.

I do not query getProgress until after the JSON entries are queued, so that is not the issue. If I override `getProgress()` in `AssetManager` with this code:

```
 (float)getLoadedAssets()/(getLoadedAssets()+getQueuedAssets());
```

my code works perfectly.  But if I comment that code out, it fails to load the final bitmap font 100% of the time on my MacBook Pro.  Debugging shows that the AssetDirectory/AssetManager thinks it is finished because toLoad == loaded,
however either the dependency queue or the loading queue is not empty.

**Comment by NathanSweet** [comment]
It's not easy to see where the problem is from the description. I haven't tried the repro, but it sounds a bit involved. If you have a proposed fix, it is likely easier to evaluate, otherwise we'll have to wait until someone has time to repro and investigate further.

**Comment by WalkerWhite** [comment]
I agree.  I have not had time to go deep into AssetManager to figure it out.  It does seem to have something to do with the way that the variables `toLoad` and `loaded` are updated with respect to dependencies (e.g. assets that depend on a secondary asset that is not already loaded first, but implicitly loaded on the request for the original).  That is why I initially felt it was a bigger problem and not just with TrueType fonts (though that is the easiest way to queue a dependent asset).

Part of me wonders it if its because my code calls load on AssetManager in an actual loading thread (e.g. I load a JSON in a JSON loader task which then queues up other tasks).  But everything appears to be thread safe and I am not doing anything that would require the main render thread context.

For now, I have been working around it by making getProgress to be the code I mentioned earlier:

```
    public synchronized float getProgress () {
        // This is a workaround for the problem with getProgress in AssetManager
        if (getLoadedAssets() == 0) { return 0.0f; }
        return (float)getLoadedAssets()/(getLoadedAssets()+getQueuedAssets());
    }
```

This causes a view hiccups if you use this to generate a loading progress bar, because queue assets counts thread tasks that do not materialize into a loaded asset.  But has the benefit that it never hits the value 1 early.

**Comment by WalkerWhite** [comment]
Looking quickly at the code for AssetManager I see something that MAY be at fault.  Dependencies are handled by `injectDependency`.  This method does not call `load`.  Instead it calls `addTask` directly without incrementing `toLoad`.  However, the variable `loaded` is updated in `updateTask` which depends on the task queue.  So you have a task in the queue which does not update `toLoad`, but does update `loaded`.  This would cause a mismatch.
